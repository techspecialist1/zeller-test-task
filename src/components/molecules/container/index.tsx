import React, {FC} from 'react';
import {SafeAreaView, View} from 'react-native';

import {styles} from './styles';

type ContainerProps = {
  children: React.ReactNode;
};

const Container: FC<ContainerProps> = ({children}) => {
  return (
    <SafeAreaView style={styles.safeArea}>
      <View style={styles.container}>{children}</View>
    </SafeAreaView>
  );
};

export default Container;
