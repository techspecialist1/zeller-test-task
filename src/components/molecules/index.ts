import ListView from './listView';
import Container from './container';
import RadioButton from '../atoms/radioButton';
import SearchInput from '../atoms/searchInput';
import ActivityIndicator from '../atoms/activityIndicator';
import ErrorBox from '../atoms/errorBox';

export {
  ListView,
  ActivityIndicator,
  RadioButton,
  SearchInput,
  Container,
  ErrorBox,
};
