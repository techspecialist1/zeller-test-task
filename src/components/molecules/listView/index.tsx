import React from 'react';
import {View, Text, TouchableOpacity} from 'react-native';

import {styles} from './style';

type ListViewProps = {
  item: {
    name: string;
    role: string;
  };
  onPress: () => void;
};

const ListView: React.FC<ListViewProps> = props => {
  const {item, onPress} = props;

  return (
    <TouchableOpacity onPress={onPress}>
      <View style={styles.userList}>
        <View style={styles.profile}>
          <Text style={styles.profileText}>{item?.name?.slice(0, 1)}</Text>
        </View>
        <View style={styles.listName}>
          <Text style={styles.name}>{item?.name}</Text>
          <Text style={styles.role}>{item?.role}</Text>
        </View>
      </View>
    </TouchableOpacity>
  );
};

export default ListView;
