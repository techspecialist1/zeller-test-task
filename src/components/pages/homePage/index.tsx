import React, {useState, useEffect, useCallback} from 'react';
import {
  View,
  Text,
  FlatList,
  RefreshControl,
  TouchableOpacity,
} from 'react-native';
import {NavigationProp} from '@react-navigation/native';

import {styles} from './style';

import {STRINGS} from '../../../utils/constants';
import {
  ActivityIndicator,
  Container,
  ListView,
  RadioButton,
  SearchInput,
  ErrorBox,
} from '../../molecules';
import {useQuery} from '@apollo/client';
import {LIST_ZELLER_ITEMS} from '../../../network/graphql/queries/queries';
import {RootStackProps} from '../../../navigation/homeStackNavigator';

type HomePageNavigationProps = NavigationProp<RootStackProps, 'Home'>;

type HomePageProps = {
  navigation: HomePageNavigationProps;
};

type ItemTypes = {
  id: string;
  role: string;
  name: string;
};

const HomePage: React.FC<HomePageProps> = ({navigation}) => {
  const {data, loading, refetch, error} = useQuery(LIST_ZELLER_ITEMS);
  const [userList, setUserList] = useState([]);
  const [searchText, setSearchText] = useState<string>('');
  const [activeRadioBtn, setActiveRadioBtn] = useState<string>(
    STRINGS.ADMIN,
  );

  const {items: users} = data?.listZellerCustomers || {};

  const handleSearchTextChange = useCallback(
    (text: string) => {
      setSearchText(text);
      const filterlist = users
        ? users?.filter((val: any) => {
            return text.length > 0
              ? val.name?.toLowerCase().includes(text.toLowerCase()) &&
                  val.role === activeRadioBtn
              : val.role === activeRadioBtn;
          })
        : [];
      setUserList(filterlist);
    },
    [users, activeRadioBtn],
  );

  // filter list after half second once user stop typing
  useEffect(() => {
    const delay = setTimeout(
      () => {
        if (activeRadioBtn || searchText) {
          handleSearchTextChange(searchText);
        }
      },
      searchText?.length ? 500 : 0,
    );
    return () => clearTimeout(delay);
  }, [activeRadioBtn, handleSearchTextChange, users, searchText]);

  const renderItem = useCallback(
    ({item}: {item: ItemTypes}) => {
      return (
        <ListView
          key={item?.id}
          item={item}
          onPress={() => navigation.navigate('Details')}
        />
      );
    },
    [navigation],
  );

  return (
    <Container>
      <ActivityIndicator isVisible={loading} />

      <View style={styles.userTypeView}>
        <Text style={styles.userTypeText}>{STRINGS.USER_TYPE}</Text>
      </View>

      <RadioButton
        label={STRINGS.ADMIN}
        onChange={(value: React.SetStateAction<string>) => {
          setActiveRadioBtn(value);
        }}
        activeButton={activeRadioBtn}
        radioSize={20}
      />
      <RadioButton
        onChange={(value: React.SetStateAction<string>) => {
          setActiveRadioBtn(value);
        }}
        radioSize={20}
        activeButton={activeRadioBtn}
        label={STRINGS.MANAGER}
      />

      <View style={styles.line} />
      <SearchInput
        placeholder={STRINGS.SEARCH_TEXT}
        onChangeText={val => handleSearchTextChange(val)}
      />
      <View style={styles.listView}>
        {activeRadioBtn && (
          <View style={styles.userNameView}>
            <Text style={styles.listUserTest}>
              {activeRadioBtn} {STRINGS.USER}
            </Text>
            {error ? (
              <TouchableOpacity onPress={() => refetch()}>
                <Text style={styles.refreshTest}>Refresh</Text>
              </TouchableOpacity>
            ) : null}
          </View>
        )}
        {error ? (
          <ErrorBox text={error?.message} />
        ) : (
          <FlatList
            showsVerticalScrollIndicator={false}
            data={userList}
            renderItem={renderItem}
            refreshControl={
              <RefreshControl
                refreshing={loading}
                onRefresh={() => refetch()}
              />
            }
          />
        )}
      </View>
    </Container>
  );
};

export default HomePage;
