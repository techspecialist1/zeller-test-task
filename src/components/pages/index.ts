import HomePage from './homePage';
import DetailPage from './detailPage';

export {HomePage, DetailPage};
