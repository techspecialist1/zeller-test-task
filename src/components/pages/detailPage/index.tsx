import React from 'react';
import {Text} from 'react-native';
import {Container} from '../../molecules';

const DetailPage = () => {
  return (
    <Container>
      <Text>Details Page</Text>
    </Container>
  );
};

export default DetailPage;
