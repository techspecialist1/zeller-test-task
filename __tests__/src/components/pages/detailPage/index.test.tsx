import React from 'react';
import renderer from 'react-test-renderer';
import {Container} from '../../../../../src/components/molecules';
import {Text} from 'react-native';
import {DetailPage} from '../../../../../src/components/pages';

describe('DetailPage', () => {
  test('renders correctly', () => {
    const component = renderer.create(<DetailPage />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders a Container component', () => {
    const component = renderer.create(<DetailPage />);
    const instance = component.root;
    const container = instance.findByType(Container);

    expect(container).toBeTruthy();
  });

  test('renders Text with "Details"', () => {
    const component = renderer.create(<DetailPage />);
    const instance = component.root;
    const text = instance.findByType(Text);

    expect(text.props.children).toBe('Details');
  });
});
