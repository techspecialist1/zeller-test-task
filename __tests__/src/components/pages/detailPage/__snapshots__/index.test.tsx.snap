// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPage renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "FFFFFF",
        "flex": 0.7,
        "padding": 19.692307692307693,
      }
    }
  >
    <Text>
      Details
    </Text>
  </View>
</RCTSafeAreaView>
`;
